### PR TITLE
Refactor filtering again: this time make subclasses of `Filter` with `filter_records(...)` methods

### LINKED ISSUE #314: Make the new `Filter` objects capable of filtering a data vector or 2d array
The new refactored code in `optimal_filter.py` and specifically the new `Filter` class objects ought to be able to, you know, _filter_ data.  See https://github.com/usnistgov/mass/pull/313#discussion_r1871739760 for this suggestion.

My concept: create subclasses of `Filter`:
1. `Filter5Lag` (includes both classic 5-lag filters and the Fourier sort)
2. `FilterATS`

Each will contain its own specific implementation of methods:
1. `filter_records(x)` (return both filtered value and phase)
2. `filter_records_nophase(x)` (same, except don't compute the phase). Drop this idea if there's no situation where it's advantageous.

In each method, the argument `x` can be a 1d or 2d array, or a numpy memmap of same. If it's 2d, then filter each row (column?) as a 1d array.

### PR DESCRIPTION
**Pull request by joefowler**
This removes filtering algorithms from the `MicrocalDataSet` into the `filter_records(...)` methods of the various subclasses of `Filter`. In so doing, many `if` branches are removed (or at least, they are hidden from the user).

Fixes #314.

### CODE REVIEW

File: mass/core/optimal_filtering.py [hunk 1]
```diff
@@ -261,12 +261,15 @@ class Filter:
...
     cut_post: int = 0
-    _filter_type: str = ""
 
     @property
     def is_arrival_time_safe(self):
         """Is this an arrival-time-safe filter?"""
-        return self.convolution_lags == 1 and self.dt_values is not None
+        return False
+
+    @property
```
Comment (ggggggggg) [inline mass/core/optimal_filtering.py:270]: Are we supposed to never make an object of type `Filter`? In that case consider naming is `BaseFilter` and `def __init__(self, ...): raise Error`? Move the error as early as possible.

File: mass/core/optimal_filtering.py [hunk 2]
```diff
@@ -296,6 +299,126 @@ def report(self, std_energy: float = 5898.8):
...
+        (10, -5, -10, -5, 10),
+    ), dtype=float) / 70.0
+
+    def filter_records(self, x: npt.ArrayLike) -> tuple[np.ndarray, np.ndarray]:
+        """Filter one microcalorimeter record or an array of records.
+
+        Parameters
+        ----------
+        x : npt.ArrayLike
+            A 1-d array, a single pulse record, or a 2-d array, each row a pulse records.
```
Comment (ggggggggg) [inline mass/core/optimal_filtering.py:334]: For those of us (me) who find it impossible to remember which index is a row, can we write "`x[i,:]` is the `i`th pulse."

File: mass/core/optimal_filtering.py [hunk 3]
```diff
@@ -296,6 +299,126 @@ def report(self, std_energy: float = 5898.8):
...
+        Raises
+        ------
+        AssertionError
+            If the input array is the wrong length
+        """
+        x = np.asarray(x)
+        if x.ndim == 1:
+            x = x.reshape((1, len(x)))
+        _, nsamp = x.shape
+        print("Heyho! ", nsamp, x.shape, len(self.values), type(self))
```
Comment (ggggggggg) [inline mass/core/optimal_filtering.py:413]: This debug comment can probably be removed.

File: mass/core/optimal_filtering.py [hunk 4]
```diff
@@ -296,6 +299,126 @@ def report(self, std_energy: float = 5898.8):
...
+
+        assert nsamp == len(self.values)
+        conv0 = np.dot(x, self.values)
+        conv1 = np.dot(x, self.dt_values)
+        ArrivalTime = conv1 / conv0
```
Comment (ggggggggg) [inline mass/core/optimal_filtering.py:418]: Capital letters conventionally refer to a class, rename `ArrivalTime` to `arrival_time` or `filt_phase` or similar. Github even shows it as orange, just like the class name `FilterMaker` and others.

**Review by ggggggggg — COMMENTED** [review]
A few small edit requests.

File: mass/core/optimal_filtering.py (hunk 1 above)
Comment (joefowler) [reply mass/core/optimal_filtering.py:270]: You're right. And...it turns out there's [a package](https://docs.python.org/3/library/abc.html) and a [special decorator](https://docs.python.org/3/library/abc.html#abc.abstractmethod) to enforce this. I'll use that.

File: mass/core/optimal_filtering.py (hunk 2 above)
Comment (joefowler) [reply mass/core/optimal_filtering.py:334]: Yes, done (in next commit).

File: mass/core/optimal_filtering.py (hunk 4 above)
Comment (joefowler) [reply mass/core/optimal_filtering.py:418]: Good point. This is _very_ old code. I guess this is relic of an earlier time.

**Pull request merged**